### PR TITLE
Refinement types

### PR DESCRIPTION
**Pull request by lorchrob**
Support refinement types

### CODE REVIEW

File: src/lustre/lustreNodeGen.ml [hunk 1]
```diff
@@ -1597,6 +1598,28 @@ and compile_node_decl gids is_function cstate ctx i ext inputs outputs locals it
...
+      let ident = mk_ident id in
+      let index_types = compile_ast_type cstate ctx map (A.Bool dummy_pos) in
+      let over_indices = fun index index_type accum ->
+        let possible_state_var = mk_state_var
+          map
+          (node_scope @ I.reserved_scope)
+          ident
+          index
+          index_type
+          (Some N.KGhost)
```
Comment (daniel-larraz) [inline src/lustre/lustreNodeGen.ml:1614]: Are all these state variables really `KGhost` variables?

**Review by daniel-larraz — CHANGES_REQUESTED** [review]
I left several comments in the PR that you need to address. The PR branch has also conflicts that you must resolve.

File: src/lustre/lustreSyntaxChecks.mli [hunk 2]
```diff
@@ -55,9 +59,21 @@ type error = [
...
+  | UnusedBoundVariableWarning of HString.t
+
+type warning = [
+  | `LustreSyntaxChecksWarning of Lib.position * warning_kind
+]
+
+val warning_message : warning_kind -> string
+
+val syntax_check : LA.t -> (([> `LustreSyntaxChecksWarning of Lib.position * warning_kind] list * LA.declaration list), 
+                            [> `LustreSyntaxChecksError of Lib.position * error_kind ]) result
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreSyntaxChecks.mli]: Addressed

File: src/lustre/lustreSyntaxChecks.ml [hunk 3]
```diff
@@ -105,6 +107,8 @@ let error_message kind = match kind with
     ^ HString.string_of_hstring node ^ "' in the cone of influence of this contract: " ^ kind ^ " "
     ^ HString.string_of_hstring node ^ " has a refinable contract"
   | NodeCallInConstant id -> "Illegal node call or 'any' operator in definition of constant '" ^ HString.string_of_hstring id ^ "'"
+  | NodeCallInGlobalTypeDecl id -> "Illegal node call or choose operator in definition of global type '" ^ HString.string_of_hstring id ^ "'"
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreSyntaxChecks.ml]: Addressed

File: src/lustre/lustreSyntaxChecks.ml [hunk 4]
```diff
@@ -996,23 +1088,26 @@ let no_mismatched_clock is_bool e =
         | _ -> syntax_error pos (IllegalClockExprInActivate c)
       in
       clocks_match_result pos clk_exp clock
-    | _ -> Ok ()
+    | _ -> Ok []
   in
-  let check_merge = function
+  let check_merge: LA.expr -> ( [> `LustreSyntaxChecksWarning of Lib.position * warning_kind ] list, [> `LustreSyntaxChecksError of Lib.position * error_kind ])
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreSyntaxChecks.ml]: Addressed

File: src/lustre/lustreTypeChecker.mli [hunk 5]
```diff
@@ -86,33 +87,58 @@ type error = [
...
+type source = 
+  | Input | Output | Local | Global | Ghost
+
 val error_message: error_kind -> string
 
 val type_error: Lib.position -> error_kind -> ('a, [> error]) result 
 (** [type_error] returns an [Error] of [tc_result] *)
      
-val type_check_infer_globals: tc_context -> LA.t -> (tc_context, [> error]) result  
+val type_check_infer_globals: tc_context -> LA.t -> (tc_context * [> `LustreTypeCheckerWarning of Lib.position * warning_kind ] list, [> error]) result
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreTypeChecker.mli]: Addressed

File: tests/regression/falsifiable/ref_type_ghost_var.lus [hunk 6]
```diff
@@ -0,0 +1,8 @@
+node N(x: int) returns (y: int);
+(*@contract
+  --!! Does adding a guarantee cause a problem if this is an imported node?
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply tests/regression/falsifiable/ref_type_ghost_var.lus]: Addressed

File: tests/regression/success/ref_type_rec_project2.lus [hunk 7]
```diff
@@ -0,0 +1,10 @@
+type Nat = subtype { n: int | n >= 0 };
+type T1 = struct { fst: Nat; snd: Nat }; 
+type T2 = struct { fst: int; snd: int };
+type R = subtype { rec: T2 | rec.snd > 5 };
+
+node M(z: R) returns (y: T1);
+let
+  --!! This example should not fail (check recursively for base record type)
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply tests/regression/success/ref_type_rec_project2.lus]: Addressed

File: src/lustre/lustreAstHelpers.mli [hunk 8]
```diff
@@ -40,6 +40,9 @@ val pos_of_expr : expr -> Lib.position
 val expr_contains_call : expr -> bool
 (** Checks if the expression contains a call to a node *)
 
+val expr_contains_id : index -> expr -> bool
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstHelpers.mli]: Addressed

File: tests/ounit/lustre/lustreTypeChecker/nondeterministic_choice.lus [hunk 9]
```diff
@@ -1,6 +1,6 @@
 node main () returns (y: int); 
 let 
-  y = any { x: bool | x < 50 };
+  y = any subtype { x: bool | x < 50 };
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply tests/ounit/lustre/lustreTypeChecker/nondeterministic_choice.lus]: Addressed

File: src/lustre/lustreAstDependencies.ml [hunk 10]
```diff
@@ -955,7 +1000,16 @@ let mk_graph_contract_node_eqn2: dependency_analysis_data -> LA.contract_node_eq
...
+    | UntypedConst (pos, i, e) -> 
+      let union g v = union_dependency_analysis_data g
+      (singleton_dependency_analysis_data empty_hs v pos)
+      in
+      let vars = vars_with_flattened_nodes ad.nsummary 0 (LH.abstract_pre_subexpressions e) in
+      let effective_vars = LA.SI.elements vars in
+      let vars2 = vars_with_flattened_nodes ad.nsummary2 0 (LH.abstract_pre_subexpressions e) in
+      let effective_vars2 = LA.SI.elements vars2 in
+      let ad = connect_g_pos_biased false (List.fold_left union ad effective_vars) i pos in
+      R.ok (connect_g_pos_biased true (List.fold_left union ad effective_vars2) i pos)
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstDependencies.ml]: Addressed (I reverted the changes)

File: src/lustre/lustreAstHelpers.ml [hunk 11]
```diff
@@ -739,6 +826,20 @@ let rec vars_of_struct_item = function
...
+    List.fold_left SI.union SI.empty (List.map vars_of_type tys)
+  | RecordType (_, _, tis) -> 
+    let vars = List.map (fun (_, _, ty) -> vars_of_type ty) tis in 
+    List.fold_left SI.union SI.empty vars
+  | RefinementType (_, (_, id, ty), e) ->
+    let vars1 = SI.diff (vars_without_node_call_ids e) (SI.singleton id) in 
+    let vars2 = vars_of_type ty in 
+    SI.union vars1 vars2
+  | Int _ | Int8 _ | Int16 _ | Int32 _ | Int64 _ | UInt8 _ | UInt16 _ | UInt32 _ | UInt64 _ | Bool _ 
+  | TVar _ | IntRange _ | Real _ | UserType _ | AbstractType _ | EnumType _ | History _ | TArr _ -> SI.empty
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstHelpers.ml]: Addressed

File: src/lustre/lustreAstHelpers.ml [hunk 12]
```diff
@@ -112,18 +143,29 @@ let rec type_contains_subrange = function
...
-let rec type_contains_enum_or_subrange = function
+let rec type_contains_ref = function
+  | RefinementType _ -> true
+  | TupleType (_, tys) | GroupType (_, tys) ->
+    List.fold_left (fun acc ty -> acc || type_contains_ref ty) false tys
+  | RecordType (_, _, tys) ->
+    List.fold_left (fun acc (_, _, ty) -> acc || type_contains_ref ty)
+      false tys
+  | ArrayType (_, (ty, _)) -> type_contains_ref ty
+  | _ -> false
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstHelpers.ml:154]: Addressed.
Note: To handle `history(x)`, we need to look up the type of `x` using a helper function from `typeCheckerContext`. But, this causes a circular dependency. So, I moved these functions (along with `type_contains_subrange`) to `typeCheckerContext` (and also added an extra argument `ctx`)

File: src/lustre/lustreAstDependencies.ml [hunk 13]
```diff
@@ -300,6 +300,7 @@ let rec mk_graph_type: LA.lustre_type -> dependency_analysis_data = function
   | ArrayType (_, (ty, e)) -> union_dependency_analysis_data (mk_graph_type ty) (mk_graph_expr e)
   | History _ -> empty_dependency_analysis_data
   | TArr (_, aty, rty) -> union_dependency_analysis_data (mk_graph_type aty) (mk_graph_type rty)
+  | RefinementType (_, _, expr) -> mk_graph_expr expr
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstDependencies.ml]: Addressed

File: src/lustre/lustreAstDependencies.ml [hunk 14]
```diff
@@ -364,6 +365,7 @@ let mk_graph_const_decl: LA.const_decl -> dependency_analysis_data
 let mk_graph_type_decl: LA.type_decl -> dependency_analysis_data
   = function
   | FreeType (pos, i) -> singleton_dependency_analysis_data ty_prefix  i pos 
+  | AliasType (pos, i, LA.RefinementType (_, (_, _, ty), _)) -> connect_g_pos (mk_graph_type ty) (HString.concat2 ty_prefix i) pos;
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreAstDependencies.ml]: Addressed. (I checked, and your conjecture is correct)

File: src/lustre/lustreTypeChecker.ml [hunk 15]
```diff
@@ -402,13 +423,48 @@ let update_ty_with_ctx node_ty call_params ctx arg_exprs =
...
+      let ty = expand_nested_type_syn ctx ty in
+      expand_type ctx ty
   )
-  | ty -> R.ok (expand_type_syn ctx ty)
+  | LA.RefinementType (_, (_, _, ty), _) -> 
+    let ty = expand_nested_type_syn ctx ty in
+    expand_type ctx ty
+  | ty -> R.ok (expand_nested_type_syn ctx ty)
+
+let rec get_base_type ctx = function
```
[the remark this answers is not in the document]
Comment (lorchrob) [reply src/lustre/lustreTypeChecker.ml]: I pushed a commit that refactors `expand_type` and related functions. It is a bit verbose and can maybe be simplified/refactored further, but the hope is that it increases clarity of the differences between the functions

File: src/lustre/lustreAstDependencies.ml [hunk 16]
```diff
@@ -420,8 +421,19 @@ let rec get_node_call_from_expr: LA.expr -> (LA.ident * Lib.position) list
...
-  = function
+let rec extract_node_calls_type: LA.lustre_type -> (LA.ident * Lib.position) list 
+= function 
+  | RefinementType (_, (_, _, ty), e) -> extract_node_calls_type ty @ get_node_call_from_expr e
+  | ArrayType (_, (ty, _)) -> extract_node_calls_type ty 
+  | TupleType (_, tys)
+  | GroupType (_, tys) -> List.map extract_node_calls_type tys |> List.flatten 
+  | RecordType (_, _, tis) -> List.map (fun (_, _, ty) -> extract_node_calls_type ty) tis |> List.flatten
+  | Int _ | Int8 _ | Int16 _ | Int32 _ | Int64 _ | UInt8 _ | UInt16 _ | UInt32 _ | UInt64 _ 
+  | Bool _ | Real _ | TVar _ | IntRange _ | UserType _ | AbstractType _ | EnumType _ | History _ | TArr _ -> []
```
Comment (daniel-larraz) [inline src/lustre/lustreAstDependencies.ml:432]: `TArr` can also include node calls.

Comment (lorchrob) [reply src/lustre/lustreAstDependencies.ml]: Addressed

**Review by daniel-larraz — APPROVED** [review]

**Pull request merged**